src/typing/type_inference_js.ml

```diff
@@ -1292,8 +1292,29 @@ and toplevels cx stmts =
       if uc < List.length stmts
       then (
         let msg = "unreachable code" in
-        let (loc, _) = List.nth stmts uc in
-        Flow_js.add_warning cx [mk_reason "" loc, msg]
+        let rec drop n lst =
+          match (n, lst) with
+          | (_, []) -> []
+          | (0, l) -> l
+          | (x, h :: t) -> drop (x-1) t
+        in
+        let trailing = drop uc stmts in
+        trailing |> List.iter Ast.Statement.(fun stmt ->
+          match stmt with
+          (* function declarations are hoisted, so not unreachable *)
+          | (_, FunctionDeclaration {
```

Comment:
With pleasure, [here](https://realworldocaml.org/v1/en/html/records.html#idm181618235488) there is an explaination about how correctly pattern match on record which is part of a module. (the syntax maybe seems ugly but there is a proposal for adding local open in pattern matching to do `Foo.{bar; toz}` instead of `{Foo.bar; toz}`, hope it will be accepted)

[ocaml.org](https://ocaml.org/learn/tutorials/) and [Real World OCaml](https://realworldocaml.org/v1/en/html/index.html) are really great places to learn OCaml, but maybe you already known these websites.

To do what you want with the `when` guard, I suggest you to use [List.for_all](http://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html#VALfor_all) with `declarations` and a predicate function.

I hope you will understand my english, don't hesite to ask me something (and come to #flowtype channel on freenode if you want).